20231018-access-logs.md

```diff
@@ -0,0 +1,188 @@
+
+### Goals
+
+- Make it possible for customers to enable access logs per bucket.
+- Collect access logs from Linksharing, Gateway-MT, and Satellite API pods.
+- Store the access logs into a target customer's Storj bucket.
+- The delay between an event happening and being recorded in the access logs should not exceed one hour.
+- Explore ways to provide real-time access logs.
+- Have a solution that works for all Storj product flavors.
+  
+### Approach / Design
+
+The design described below is based on [Grafana Loki](https://grafana.com/oss/loki/) - a log aggregation stack by Grafana.
+
+Grafana Loki comes in a few different flavors:
+- [Open source stack](https://grafana.com/oss/loki/) (AGPLv3) to use in the Storj infrastructure to aggregate and push logs to customer's Storj buckets.
+- [Grafana Cloud](https://grafana.com/products/cloud/logs/) - a SaaS solution that we can use for providing real-time access logs.
+- [Grafana Enterprise Logs (GEL)](https://grafana.com/docs/enterprise-logs/latest/) - a commercial on-premise solution that we can use for large Storj Private Cloud installations.
+
+#### Data fields in the access logs
+
+Below are the data fields to include in the access logs.
+
+We should avoid logging any secrets like access grants or sensitive data like individual IP addresses. See the [Security / Privacy](#security--privacy) section for mode details.
+
+| Name | Description | Sample Value | Notes |
+|------|-------------|--------------|-------|
+| Timestamp | The time at which the request was received; these dates and times are in Coordinated Universal Time (UTC), following the ISO 8601 standard. | 2023-10-18T09:37:06Z |  |
+| HTTP Method | The HTTP method used for the request (HEAD, GET). | GET | Amazon's S3 Bucket Logs return something called "Operation" which looks like `REST.PUT.OBJECT` |
+| Referer | The value of the HTTP Referer header, if present. | http://www.example.com/webservices |  |
+| User Agent | The value of the HTTP User-Agent header. | curl/7.15.1 |  |
+| Object Size | The total size of the object, in bytes. | 3462992 |  |
+| Bytes Sent | The number of response bytes sent, excluding the HTTP protocol overhead, or - if zero. | 2662992 |  |
+| HTTP Status | The numeric HTTP status code of the response. | 200 |  |
+| Project | The hex-encoded Public ID of the Storj project that is serving the request. | ef1b109249684b3e8466a865ce22a262 |  |
+| Bucket | The bucket name that is serving the request. | test-bucket |  |
+| Object Key | The object key (path) that is requested. | /photos/IMG_20200301_153448733.jpg |  |
+| Access Key ID | The S3 access key ID used to authenticate the request. | juudmr5vp28bjrqhny7luyibs9pq |  |
+| Macaroon Head | The hex-encoded macaroon head of the API key. | 70d745b9cb906bb75b4979a8160b8e8bef8e9d6b10d64d6ecd0f17efa02d649e |  |
+| Satellite Address | The Node URL of the satellite serving the request. | 12EayRS2V1kEsWESU9QMRseFhdxYxKicsiFmxrsLZHeLUtdps3S@us1.storj.io:7777 |  |
```

Comment:
isn't this going to be identical for every single log line in the bucket?  Doesn't seem worth logging.

Reply:
Yes. Same for `Project` and `Bucket.` However, it may help to quickly identify the satellite just by looking at the logs. I prefer to be verbose especially on data that we can easily provide. We may reconsider to remove it later.